workspace_tools/upload_results.py

```diff
@@ -0,0 +1,171 @@
+
+import pprint
+
+import requests
+import urlparse
+
+pp = pprint.PrettyPrinter(indent=4)
+
+def create_headers(args):
+    return { 'X-Api-Key':  args.api_key }
+
+def finish_command(command, response):
+    print(command, response.status_code, response.reason)
+    print(response.text)
+
+    if response.status_code < 400:
+        sys.exit(0)
+    else:
+        sys.exit(2)
+
+def create_build(args):
+    build = {}
+    build['buildType'] = args.build_type
+    build['number'] = args.build_number
+    build['source'] = args.build_source
+    build['status'] = 'running'
+
+    if build['buildType'] == 'Pull_Request':
+        build['buildType'] = 'Pull Request'
+
+    r = requests.post(urlparse.urljoin(args.url, "api/builds"), headers=create_headers(args), json=build)
+
+    if r.status_code < 400:
+        if args.property_file_format:
+            print("MBED_BUILD_ID=" + r.text)
+        else:
+            print(r.text)
+
+        sys.exit(0)
+    else:
+        sys.exit(2)
+
+def finish_build(args):
+    data = {}
+    data['status'] = 'completed'
+
+    r = requests.put(urlparse.urljoin(args.url, "api/builds/" + args.build_id), headers=create_headers(args), json=data)
+    finish_command('finish-build', r)
+
+def abort_build(args):
+    data = {}
+    data['status'] = 'aborted'
+
+    r = requests.put(urlparse.urljoin(args.url, "api/builds/" + args.build_id), headers=create_headers(args), json=data)
+    finish_command('abort-build', r)
+
+def add_test_runs(args):
+    tree = None
+
+    try:
+        tree = ET.parse(args.test_report)
+    except:
+        print(sys.exc_info()[0])
+        print('Invalid path to test report.')
+        sys.exit(1)
+
+    test_suites = tree.getroot()
+
+    ts_data = {}
+    ts_data['testRuns'] = []
+
+    platforms_set = set()
+    toolchains_set = set()
+    testIds_set = set()
+    hostOses_set = set()
+
+    hostOses_set.add(args.host_os)
+
+    for test_suite in test_suites:
+        platform = ""
+        toolchain = ""
+        for properties in test_suite.findall('properties'):
+            for property in properties.findall('property'):
+                if property.attrib['name'] == 'target':
+                    platform = property.attrib['value']
+                    platforms_set.add(platform)
+                elif property.attrib['name'] == 'toolchain':
+                    toolchain = property.attrib['value']
+                    toolchains_set.add(toolchain)
+
+        for test_case in test_suite.findall('testcase'):
+            testRun = {}
+            testRun['build'] = args.build_id
+            testRun['hostOs'] = args.host_os
+            testRun['platform'] = platform
+            testRun['toolchain'] = toolchain
+            testRun['test'] = test_case.attrib['classname'].split('.')[-1]
+
+            testIds_set.add(testRun['test'])
+
+            system_outs = test_case.findall('system-out')
+
+            if system_outs:
+                testRun['output'] = system_outs[0].text
```

Comment:
Based on the sample xml test reports I went through, there is only one "system-out" element per "testcase" element. So I always just grab the first (and only) output text.